processing/src/main/java/org/apache/druid/query/aggregation/firstlast/FirstLastVectorAggregator.java

```diff
@@ -0,0 +1,271 @@
+   *
+   *  timeSelector can be null, however aggregate functions are no-op then. Weird, since all numeric versions supply the selector,
+   *  only the case when the aggregator's capabilities are not present in the string version do we hit this case (not sure why this is
+   *  a possibility, and what benefit does it provides)
+   */
+  public FirstLastVectorAggregator(
+      @Nullable VectorValueSelector timeSelector,
+      @Nullable VectorValueSelector valueSelector,
+      @Nullable VectorObjectSelector objectSelector,
+      SelectionPredicate selectionPredicate
+  )
+  {
+    if (timeSelector != null) {
+      Preconditions.checkArgument(
+          (valueSelector != null && objectSelector == null) || (valueSelector == null && objectSelector != null),
+          "exactly one of 'valueSelector' and 'objectSelector' must be provided"
+      );
+    }
+    this.timeSelector = timeSelector;
+    this.valueSelector = valueSelector;
+    this.objectSelector = objectSelector;
+    this.selectionPredicate = selectionPredicate;
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position, int startRow, int endRow)
+  {
+    // Not a normal case, and this doesn't affect the folding. timeSelectors should be present (albeit irrelevent) when folding.
+    // timeSelector == null means that the aggregating column's capabilities aren't known, and it only happens for a special case
+    // while building string aggregator
+    if (timeSelector == null) {
+      return;
+    }
+
+    // If objectSelector isn't null, then the objects might be folded up. If that's the case, whatever's represented by
+    // the timeSelector doesn't hold any relevance.
+    if (objectSelector != null) {
+      final Object[] maybeFoldedObjects = objectSelector.getObjectVector();
+      final boolean[] timeNullityVector = timeSelector.getNullVector();
+      final long[] timeVector = timeSelector.getLongVector();
+
+      PairType selectedPair = null;
+
+      for (int index = startRow; index < endRow; ++index) {
+
+        PairType pair = readPairFromVectorSelectors(timeNullityVector, timeVector, maybeFoldedObjects, index);
+        if (pair != null) {
+          if (selectedPair == null) {
+            selectedPair = pair;
+          } else if (selectionPredicate.apply(pair.lhs, selectedPair.lhs)) {
+            selectedPair = pair;
+          }
+        }
+      }
+      // Something's been selected of the row vector
+      if (selectedPair != null) {
+        // Compare the latest value of the folded up row vector to the latest value in the buffer
+        if (selectionPredicate.apply(selectedPair.lhs, buf.getLong(position))) {
+          if (selectedPair.rhs != null) {
+            putValue(buf, position, selectedPair.lhs, selectedPair.rhs);
+          } else if (useDefault) {
+            putDefaultValue(buf, position, selectedPair.lhs);
+          } else {
+            putNull(buf, position, selectedPair.lhs);
+          }
```

Comment:
I kept it separate to prevent autoboxing in the `putDefaultValue`. The callers can directly put the default value, without needing to convert it from value to Object and then back to value. `putNull` can probably be removed.